File: fudi-nfc/src/main/java/com/romellfudi/fudinfc/util/sync/WriteUtilityImpl.java
package com.romellfudi.fudinfc.util.sync;

import android.nfc.FormatException;
import android.nfc.NdefMessage;
import android.nfc.Tag;
import android.nfc.tech.Ndef;
import android.nfc.tech.NdefFormatable;
import android.util.Log;

import com.romellfudi.fudinfc.util.exceptions.InsufficientCapacityException;
import com.romellfudi.fudinfc.util.exceptions.ReadOnlyTagException;
import com.romellfudi.fudinfc.util.interfaces.NdefWrite;
import com.romellfudi.fudinfc.util.interfaces.WriteUtility;


public class WriteUtilityImpl implements WriteUtility {

    private static final String TAG = WriteUtilityImpl.class.getName();

    private boolean readOnly = false;

    private NdefWrite mNdefWrite;

    public WriteUtilityImpl() {
        setNdefWrite(new NdefWriteImpl());
    }

    /**
     * @param ndefWrite
     *         used to delegate writing to Tag
     *
     * @throws java.lang.NullPointerException
     *         when null
     */
    public WriteUtilityImpl(NdefWrite ndefWrite) {
        if (ndefWrite == null) {
            throw new NullPointerException("WriteUtility cannot be null");
        }
        setNdefWrite(ndefWrite);
    }

    private void setNdefWrite(NdefWrite ndefWrite) {
        this.mNdefWrite = ndefWrite;
    }

    @Override
    public boolean writeToNdef(NdefMessage message, Ndef ndef) throws ReadOnlyTagException, InsufficientCapacityException, FormatException {
        return mNdefWrite.writeToNdef(message, ndef);
    }

    @Override
    public boolean writeToNdefAndMakeReadonly(NdefMessage message, Ndef ndef) throws ReadOnlyTagException, InsufficientCapacityException, FormatException {
        return mNdefWrite.writeToNdefAndMakeReadonly(message, ndef);
    }

    @Override
    public boolean writeToNdefFormatable(NdefMessage message, NdefFormatable ndefFormatable) throws FormatException {
        return mNdefWrite.writeToNdefFormatable(message, ndefFormatable);
    }

    @Override
    public boolean writeToNdefFormatableAndMakeReadonly(NdefMessage message, NdefFormatable ndefFormat) throws FormatException {
        return mNdefWrite.writeToNdefFormatableAndMakeReadonly(message, ndefFormat);
    }

    @Override
    public boolean writeSafelyToTag(NdefMessage message, Tag tag) {
        try {
            writeToTag(message, tag);
        } catch (ReadOnlyTagException e) {
            Log.d(TAG, "Tag is Read only !", e);
        } catch (InsufficientCapacityException e) {
            Log.d(TAG, "The tag's capacity is insufficient!", e);
        } catch (FormatException e) {
            Log.d(TAG, "The message is malformed!", e);
        }

        return false;
    }

    @Override
    public boolean writeToTag(NdefMessage message, Tag tag) throws FormatException, ReadOnlyTagException, InsufficientCapacityException {
        Ndef ndef = Ndef.get(tag);
        NdefFormatable formatable = NdefFormatable.get(tag);

        boolean result;
        if (readOnly) {
            result = writeToNdefAndMakeReadonly(message, ndef) || writeToNdefFormatableAndMakeReadonly(message, formatable);
        } else {
            result = writeToNdef(message, ndef) || writeToNdefFormatable(message, formatable);
        }

        readOnly = false;
        return result;
    }

    public WriteUtility makeOperationReadOnly() {
        readOnly = true;
        return this;
    }
}

File: fudi-nfc/src/main/java/com/romellfudi/fudinfc/util/sync/WriteUtilityImpl.kt
/*
 * Copyright (c) 2020. BoostTag E.I.R.L. Romell D.Z.
 * All rights reserved
 * porfile.romellfudi.com
 */
package com.romellfudi.fudinfc.util.sync

import android.nfc.FormatException
import android.nfc.NdefMessage
import android.nfc.Tag
import android.nfc.tech.Ndef
import android.nfc.tech.NdefFormatable
import android.util.Log
import com.romellfudi.fudinfc.util.exceptions.InsufficientCapacityException
import com.romellfudi.fudinfc.util.exceptions.ReadOnlyTagException
import com.romellfudi.fudinfc.util.interfaces.NdefWrite
import com.romellfudi.fudinfc.util.interfaces.WriteUtility

class WriteUtilityImpl : WriteUtility {
    private var readOnly = false
    private var mNdefWrite: NdefWrite? = null

    constructor() {
        setNdefWrite(NdefWriteImpl())
    }

    /**
     * @param ndefWrite
     * used to delegate writing to Tag
     *
     * @throws java.lang.NullPointerException
     * when null
     */
    constructor(ndefWrite: NdefWrite?) {
        if (ndefWrite == null) {
            throw NullPointerException("WriteUtility cannot be null")
        }
        setNdefWrite(ndefWrite)
    }

    private fun setNdefWrite(ndefWrite: NdefWrite) {
        mNdefWrite = ndefWrite
    }

    @Throws(
        ReadOnlyTagException::class,
        InsufficientCapacityException::class,
        FormatException::class
    )
    override fun writeToNdef(message: NdefMessage?, ndef: Ndef?): Boolean {
        return mNdefWrite?.writeToNdef(message, ndef) == true
    }

    @Throws(
        ReadOnlyTagException::class,
        InsufficientCapacityException::class,
        FormatException::class
    )
    override fun writeToNdefAndMakeReadonly(message: NdefMessage?, ndef: Ndef?): Boolean {
        return mNdefWrite?.writeToNdefAndMakeReadonly(message, ndef) == true
    }

    @Throws(FormatException::class)
    override fun writeToNdefFormatable(
        message: NdefMessage?,
        ndefFormatable: NdefFormatable?
    ): Boolean {
        return mNdefWrite?.writeToNdefFormatable(message, ndefFormatable) == true
    }

    @Throws(FormatException::class)
    override fun writeToNdefFormatableAndMakeReadonly(
        message: NdefMessage?,
        ndefFormat: NdefFormatable?
    ): Boolean {
        return mNdefWrite?.writeToNdefFormatableAndMakeReadonly(message, ndefFormat) == true
    }

    override fun writeSafelyToTag(message: NdefMessage?, tag: Tag?): Boolean {
        try {
            writeToTag(message, tag)
        } catch (e: ReadOnlyTagException) {
            Log.d(TAG, "Tag is Read only !", e)
        } catch (e: InsufficientCapacityException) {
            Log.d(TAG, "The tag's capacity is insufficient!", e)
        } catch (e: FormatException) {
            Log.d(TAG, "The message is malformed!", e)
        }
        return false
    }

    @Throws(
        FormatException::class,
        ReadOnlyTagException::class,
        InsufficientCapacityException::class
    )
    override fun writeToTag(message: NdefMessage?, tag: Tag?): Boolean {
        val ndef = Ndef.get(tag)
        val formatable = NdefFormatable.get(tag)
        val result: Boolean = if (readOnly) {
            writeToNdefAndMakeReadonly(message, ndef) ||
                    writeToNdefFormatableAndMakeReadonly(message, formatable)
        } else {
            writeToNdef(message, ndef) || writeToNdefFormatable(message, formatable)
        }
        readOnly = false
        return result
    }

    override fun makeOperationReadOnly(): WriteUtility? {
        readOnly = true
        return this
    }

    companion object {
        private val TAG = WriteUtilityImpl::class.java.name
    }
}